pkg/edition/java/lite/forward.go

```diff
@@ -157,27 +170,155 @@ func findRoute(
+		backends := slices.Clone(tryBackends)
+		for len(backends) > 0 {
+			randIndex := r.Intn(len(backends))
+			backend := backends[randIndex]
+			if checkBackend(backend) {
+				return backend, log, true
+			}
+
+			backends = slices.Delete(backends, randIndex, randIndex+1)
+		}
+
+		// no working backend found
+		return "", log, false
+	}
+}
+
+var roundRobinIndex sync.Map
+
+func roundRobinNextBackend(log logr.Logger, routeHost string, tryBackends []string) nextBackendFunc {
+	return func() (string, logr.Logger, bool) {
+		if len(tryBackends) == 0 {
 			return "", log, false
 		}
-		backendAddr := dstAddr.String()
-		if _, port := netutil.HostPort(dstAddr); port == 0 {
-			backendAddr = net.JoinHostPort(dstAddr.String(), "25565")
+
+		for range tryBackends {
+			value, _ := roundRobinIndex.LoadOrStore(routeHost, 0)
+			index := value.(int)
+
+			backend := tryBackends[index%len(tryBackends)]
+			roundRobinIndex.Store(routeHost, index+1)
+
+			if checkBackend(backend) {
+				return backend, log, true
+			}
 		}
 
-		return backendAddr, log.WithValues("backendAddr", backendAddr), true
+		// no working backend found
+		return "", log, false
 	}
+}
 
-	return log, src, route, nextBackend, nil
+var leastConnectionCounter sync.Map
+
+func leastConnectionsNextBackend(log logr.Logger, tryBackends []string) nextBackendFunc {
+	return func() (string, logr.Logger, bool) {
+		if len(tryBackends) == 0 {
+			return "", log, false
+		}
+
+		var leastBackend string
+		var leastCount int = math.MaxInt32
+
+		for _, backend := range tryBackends {
+			if !checkBackend(backend) {
+				continue
+			}
+
+			count, _ := leastConnectionCounter.LoadOrStore(backend, 0)
+			if count.(int) < leastCount {
+				leastBackend = backend
+				leastCount = count.(int)
+			}
+		}
+
+		// no working background found
+		if leastBackend == "" {
+			return "", log, false
+		}
+
+		return leastBackend, log, true
+	}
+}
+
+var latencyCache = ttlcache.New[string, time.Duration]()
+
+func lowestLatencyNextBackend(log logr.Logger, tryBackends []string) nextBackendFunc {
+	return func() (string, logr.Logger, bool) {
+		if len(tryBackends) == 0 {
+			return "", log, false
+		}
+		var lowestBackend string
+		var lowestLatency time.Duration
+		for _, backend := range tryBackends {
+			if !checkBackend(backend) {
+				continue
+			}
+
+			latencyItem := latencyCache.Get(backend)
+			if latencyItem == nil {
+				latency := measureLatency(backend)
+				latencyCache.Set(backend, latency, time.Minute)
+				latencyItem = latencyCache.Get(backend)
+			}
+			if latencyItem != nil && (lowestLatency == 0 || latencyItem.Value() < lowestLatency) {
+				lowestBackend = backend
+				lowestLatency = latencyItem.Value()
+			}
+		}
+
+		// no working background found
+		if lowestBackend == "" {
+			return "", log, false
+		}
+
+		return lowestBackend, log, true
+	}
+}
+
+func measureLatency(backend string) time.Duration {
```

Comment:
We could reuse the dials that are already measure done to track latency . No need to do it separately

Reply:
Sorry I meant that we could extend the existing dial calls to measure latency of the dial.